### ISSUE TITLE
customization of c++/qml modules

### ISSUE THREAD

**Issue by udeved** [calamares/calamares#2265]
**Describe the bug**
Not directly a bug, rather than a question how to best customize eg the packagechooser(q) modules?
The calamares-extensions repo is currently broke with calamares 3.3.0 and later cmake.

If I put my customized packagechooser module in the extensions repo, the build fails with

```
CMake Error at modules/packagechooser/CMakeLists.txt:62 (calamares_add_test):
  Unknown CMake command "calamares_add_test".
```

If I comment the test, the build still fails at not finding compat/Variant.h.
Is this related to 3.3.0 changes? Or should the packagechosser be built with main package and patches as I do currently to get an image into its source?

**Comment by dalto8** [comment]
If you want a customized version of an existing module, I think many distros are using a patch in their package building process to apply the code customizations.  Alternatively, some distros maintain a fork of Calamares.

**Comment by udeved** [comment]
We have all that, but a binary patch is a PITA to get images in the source.
https://gitea.artixlinux.org/artix/calamares

Anyhow, I managed to get the packagechooser to build with the extensions after adding a patch in cmake extensions, calamares would have to install libcalamares/compat and packages headers (again?)

https://gitea.artixlinux.org/packages/calamares/src/branch/master/PKGBUILD#L31

**Comment by adriaandegroot** [comment]
- *packagechooser* is part of the main repository, so I'm not sure what -extensions has to do with any of this?
- customizing *packagechooser* in what way? The configuration file is pretty extensive, so what are you missing?
- customizing *packagechooserq* in what way? You can edit the QML file and place it in the search path, no need to rebuild anything, and configuration is just like *packagechooser* (e.g. images for selections can live in the filesystem)

I haven't tried to build -extensions at all since 3.3.0, it's quite possible that the necessary CMake stuff isn't being installed. That said there's a long-standing issue in this repo to introduce a proper -dev package from CMake.

**Comment by adriaandegroot** [comment]
By the way, the weird-ass hack in your `PKGBUILD` `prepare()` is **exactly** the kind of thing you should be reporting upstream, not papering over downstream. You click "new issue" and write "Not all headers from libcalamares are being installed" and explain that those two new (3.3.0) directories aren't being installed. Or you file a PR with the same meaning.

**Comment by udeved** [comment]
@adriaandegroot 

Happy new year.
Yes, it was a hack to just test the extension build. Question, is the compat header being ported away at some point?
Most of the default c++ modules depend on the missing headers if they are put in extensions repo.
I can send a PR https://gitea.artixlinux.org/artix/calamares/src/branch/headers-install-fix

Regarding customizing packagechooser(q) modules, I found it a pita to maintain a fork with module branches patching the modules. I added images to both packagechooser modules, and it terms of deployment, the extensions repo totally fits our purpose. :)
https://gitea.artixlinux.org/artix/calamares-extensions/src/branch/artix/modules
